### PR TITLE
NIrbhay/  Traffic mirror Session

### PR DESCRIPTION
**Pull request by Nirbhay1997**
Signed-off-by: nirbhay1997 <nkumar@progress.com>

### Description
Added Support  for AWS EC2 Traffic Mirror Filter 

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [x] New functionality includes integration tests/controls
- [x] New Terraform resources
- [x] Documentation provided or updated for resources 
- [x] All Integration Tests pass
- [x] All Unit Tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: libraries/aws_ec2_traffic_mirror_session.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require 'aws_backend'
+
+class AWSEc2TrafficMirrorSession < AwsResourceBase
+  name 'aws_ec2_traffic_mirror_session'
+  desc 'Audits EC2 Traffic Mirror Session'
+
+  example "
+    describe aws_ec2_traffic_mirror_session(traffic_mirror_session_id: 'test-traffic_mirror_session_id') do
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_session.rb:10]: Change the value to "TrafficMirrorSessionId"

File: libraries/aws_ec2_traffic_mirror_session.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  def initialize(opts = {})
+    opts = { traffic_mirror_session_id: opts } if opts.is_a?(String)
+    super(opts)
+    validate_parameters(required: %i(traffic_mirror_session_id))
+    if opts[:traffic_mirror_session_id] && !opts[:traffic_mirror_session_id].empty? # Use instance_id, if provided
+      if !opts[:traffic_mirror_session_id].is_a?(String) || opts[:traffic_mirror_session_id] !~ /(^tms-[0-9a-f]{8})|(^i-[0-9a-f]{17})$/
+        raise ArgumentError, "#{@__resource_name__}: `traffic_mirror_session_id` must be a string in the format of 'tms-' followed by 8 or 17 hexadecimal characters."
+      end
+      @display_name = opts[:traffic_mirror_session_id]
+      traffic_mirror_session_arguments = { traffic_mirror_session_ids: [opts[:traffic_mirror_session_id]] }
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_session.rb:24]: traffic_mirror_session_arguments = { traffic_mirror_session_ids: [@display_name] }

File: libraries/aws_ec2_traffic_mirror_session.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      create_resource_methods(@traffic_mirror_session)
+    end
+  end
+
+  def exists?
+    !@traffic_mirror_session.nil? && !@traffic_mirror_session.empty?
+  end
+
+  def to_s
+    "EC2 Traffic Mirror Session : #{@display_name}"
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_session.rb:38]: "EC2 Traffic Mirror Session ID: #{@display_name}"

File: test/unit/resources/aws_ec2_traffic_mirror_session_test.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  end
+end
+
+class AwsEc2TrafficMirrorSessionPathTest < Minitest::Test
+
+  def setup
+    data = {}
+    data[:method] = :describe_traffic_mirror_sessions
+    mock_lt = {}
+    mock_lt[:traffic_mirror_session_id] = 'tms-01a6e9ac9f962f154'
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_session_test.rb:29]: Change the values to dummy values.

File: test/unit/resources/aws_ec2_traffic_mirror_sessions_test.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    mock_lt[:traffic_mirror_filter_id] = 'test-account'
+    mock_lt[:network_interface_id] = 'dummy_id'
+    mock_lt[:owner_id] = 'dummy_id'
+    data[:data] = { :traffic_mirror_sessions => [mock_lt] }
+    data[:client] = Aws::EC2::Client
+    @addr = AWSEc2TrafficMirrorSessions.new(client_args: { stub_responses: true }, stub_data: [data])
+  end
+
+  def test_aws_ec2_traffic_mirror_sessions_exists
+    require 'byebug'
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_sessions_test.rb:40]: Please remove this.

**Review by soumyo13 — CHANGES_REQUESTED** [review]
Please add the documentation and the integration tests.

File: libraries/aws_ec2_traffic_mirror_sessions.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    loop do
+      catch_aws_errors do
+        @api_response = @aws.compute_client.describe_traffic_mirror_sessions(@pagination_options)
+
+      end
+      return traffic_mirror_sessions_rows if !@api_response || @api_response.empty?
+
+      @api_response.traffic_mirror_sessions.each do |traffic_mirror_sessions|
+        @traffic_mirror_sessions_rows += [{
+
```
[the remark this answers is not in the document]
Comment (Nirbhay1997) [reply libraries/aws_ec2_traffic_mirror_sessions.rb:49]: thanks a lot

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
+---
+title: About the aws_ec2_traffic_mirror_sessions Resource
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:2]: aws_ec2_traffic_mirror_session

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+platform: aws
+---
+
+# aws\_ec2\_traffic_\mirror_\session
+
+Use the `aws_ec2_traffic_mirror_sessions` InSpec audit resource to test properties an AWS Traffic Mirror Session
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:8]: aws_ec2_traffic_mirror_session

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+## Syntax
+
+An `aws_ec2_traffic_mirror_session` resource block declares the tests for an  AWS Traffic Mirror Sessions
+
+## Syntax
+
+An `aws_ec2_launch_template` resource block declares the tests for a single  AWS Traffic Mirror Session by traffic_mirror_session_id.
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:16]: aws_ec2_traffic_mirror_session

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+platform: aws
+---
+
+# aws\_ec2\_traffic_\mirror_\session
+
+Use the `aws_ec2_traffic_mirror_sessions` InSpec audit resource to test properties an AWS Traffic Mirror Session
+
+## Syntax
+
+An `aws_ec2_traffic_mirror_session` resource block declares the tests for an  AWS Traffic Mirror Sessions
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:12]: AWS Traffic Mirror Session

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+|packet_length                | The number of bytes in each packet to mirror. |
+|virtual_network_id           | The virtual network ID associated with the Traffic Mirror session. |
+|description                  | The description of the Traffic Mirror session. |
+|tags                         | The tags assigned to the Traffic Mirror session., e.g, `[{:key=>"Name", :value=>"Testing Box"}, {:key=>"Environment", :value=>"Dev"}]`|
+
+There are also additional properties available. For a comprehensive list, see [the API reference documentation](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Instance.html)
+
+## Examples
+
+##### Test that an should exist
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:49]: Change the description

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    describe aws_ec2_traffic_mirror_session(traffic_mirror_session_id: 'test-traffic_mirror_session_id') do
+      it { should exist }
+    end
+
+##### Test that an EC2 instance is using the correct AMI
+    describe aws_ec2_traffic_mirror_session(traffic_mirror_session_id: 'test-traffic_mirror_session_id') do
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:55]: Change the description

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+      its('packet_length') { should eq 1 }
+    end
+
+##### Test that an EC2 instance has the correct tag
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:59]: Change the description

File: docs/resources/aws_ec2_traffic_mirror_sessions.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+There are also additional properties available. For a comprehensive list, see [the API reference documentation](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Instance.html)
+
+## Examples
+
+##### Test that an AWS Traffic Mirror Session should exist
+describe aws_ec2_traffic_mirror_sessions do
+  it { should exist }
+end
+
+##### Test that Traffic  mirror target includes a value
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_sessions.md:49]: Change the description

File: docs/resources/aws_ec2_traffic_mirror_sessions.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+  its('traffic_mirror_target_id') { should include 'test-mirror-id' }
+end
+
+##### Test that description is present
+describe aws_ec2_traffic_mirror_sessions do
+  its('description') { should include 'test-lt' }
+end
+
+
+## Matchers
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_sessions.md:60]: Add matchers.

File: libraries/aws_ec2_traffic_mirror_session.rb (hunk 3 above)
Comment (soumyo13) [reply libraries/aws_ec2_traffic_mirror_session.rb:38]: "EC2 Traffic Mirror Session ID: #{@display_name}"

File: libraries/aws_ec2_traffic_mirror_sessions.rb [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+             .register_column(:owner_ids, field: :owner_id)
+             .register_column(:packet_length, field: :packet_length)
+             .register_column(:session_number, field: :session_number)
+             .register_column(:virtual_network_ids, field: :virtual_network_id)
+             .register_column(:tags, field: :tags)
+             .install_filter_methods_on_resource(self, :table)
+
+  def initialize(opts = {})
+    super(opts)
+
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_sessions.rb:31]: Remove the line.

File: libraries/aws_ec2_traffic_mirror_sessions.rb [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    @traffic_mirror_sessions_rows = []
+    @pagination_options = {}
+    @pagination_options[:max_results] = 100
+    loop do
+      catch_aws_errors do
+        @api_response = @aws.compute_client.describe_traffic_mirror_sessions(@pagination_options)
+
+      end
+      return traffic_mirror_sessions_rows if !@api_response || @api_response.empty?
+
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_sessions.rb:46]: remove the line.

File: libraries/aws_ec2_traffic_mirror_sessions.rb [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+          traffic_mirror_target_id: traffic_mirror_sessions.traffic_mirror_target_id,
+          network_interface_id: traffic_mirror_sessions.network_interface_id,
+          traffic_mirror_filter_id: traffic_mirror_sessions.traffic_mirror_filter_id,
+          owner_id: traffic_mirror_sessions.owner_id,
+          packet_length: traffic_mirror_sessions.packet_length,
+          session_number: traffic_mirror_sessions.session_number,
+          virtual_network_id: traffic_mirror_sessions.virtual_network_id,
+          traffic_mirror_session_id: traffic_mirror_sessions.traffic_mirror_session_id,
+          tags: traffic_mirror_sessions.tags,
+
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_sessions.rb:60]: remove the extra line.

File: libraries/aws_ec2_traffic_mirror_sessions.rb [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+  def initialize(opts = {})
+    super(opts)
+
+    validate_parameters
+    @table = fetch_data
+  end
+
+  def fetch_data
+    @traffic_mirror_sessions_rows = []
+    @pagination_options = {}
```
Comment (soumyo13) [inline libraries/aws_ec2_traffic_mirror_sessions.rb:38]: pagination_options - we can use this.

File: test/unit/resources/aws_ec2_traffic_mirror_session_test.rb [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    mock_lt[:traffic_mirror_target_id] = 'tmt-01a6e9ac9f962f154'
+    mock_lt[:traffic_mirror_filter_id] = 'test-account'
+    mock_lt[:network_interface_id] = 'dummy_id'
+    mock_lt[:owner_id] = 'dummy_id'
+    data[:data] = { :traffic_mirror_sessions => [mock_lt] }
+    data[:client] = Aws::EC2::Client
+    @addr = AWSEc2TrafficMirrorSession.new(traffic_mirror_session_id: 'tms-01a6e9ac9f962f154',client_args: { stub_responses: true }, stub_data: [data])
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_session_test.rb:36]: Change the var name.

File: test/unit/resources/aws_ec2_traffic_mirror_session_test.rb [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+end
+
+class AwsEc2TrafficMirrorSessionPathTest < Minitest::Test
+
+  def setup
+    data = {}
+    data[:method] = :describe_traffic_mirror_sessions
+    mock_lt = {}
+    mock_lt[:traffic_mirror_session_id] = 'tms-01a6e9ac9f962f154'
+    mock_lt[:traffic_mirror_target_id] = 'tmt-01a6e9ac9f962f154'
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_session_test.rb:30]: tmt-01a6e9ac9f962f154 should be tmt-1234567890

File: test/unit/resources/aws_ec2_traffic_mirror_sessions_test.rb [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    data[:method] = :describe_traffic_mirror_sessions
+    mock_lt = {}
+    mock_lt[:traffic_mirror_session_id] = 'tms-01a6e9ac9f962f154'
+    mock_lt[:traffic_mirror_target_id] = 'tmt-01a6e9ac9f962f154'
+    mock_lt[:traffic_mirror_filter_id] = 'test-account'
+    mock_lt[:network_interface_id] = 'dummy_id'
+    mock_lt[:owner_id] = 'dummy_id'
+    data[:data] = { :traffic_mirror_sessions => [mock_lt] }
+    data[:client] = Aws::EC2::Client
+    @addr = AWSEc2TrafficMirrorSessions.new(client_args: { stub_responses: true }, stub_data: [data])
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_sessions_test.rb:36]: Same here

File: test/unit/resources/aws_ec2_traffic_mirror_sessions_test.rb [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+  end
+end
+
+class AwsEc2TrafficMirrorSessionsPathTest < Minitest::Test
+
+  def setup
+    data = {}
+    data[:method] = :describe_traffic_mirror_sessions
+    mock_lt = {}
+    mock_lt[:traffic_mirror_session_id] = 'tms-01a6e9ac9f962f154'
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_sessions_test.rb:29]: Change the hardcoded values.

**Review by soumyo13 — CHANGES_REQUESTED** [review]
Please review comments.

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+title: About the aws_ec2_traffic_mirror_sessions 
+platform: aws
+---
+
+# aws\_ec2\_traffic_\mirror_\session
+
+Use the `aws_ec2_traffic_mirror_session` InSpec audit resource to test properties an AWS Traffic Mirror Session
+
+## Syntax
+
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:11]: End the line

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+    describe aws_ec2_traffic_mirror_session(traffic_mirror_session_id: 'test-traffic_mirror_session_id') do
+      it { should exist }
+    end
+
+#### Parameters
+
+traffic_mirror_session_id must be be provided.
+
+##### traffic\_mirror_\session_\id
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:26]: Remove line no. 24. and make it as required.

File: docs/resources/aws_ec2_traffic_mirror_sessions.md [hunk 26]
```diff
@@ -0,0 +1,62 @@
+---
+title: About the aws_ec2_traffic_mirror_sessions Resource
+platform: aws
+---
+
+# aws\_ec2\_traffic_\mirror_\sessions
+
+Use the `aws_ec2_traffic_mirror_sessions` InSpec audit resource to test properties of a all the AWS Traffic Mirror Sessions
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_sessions.md:8]:
```suggestion
Use the `aws_ec2_traffic_mirror_sessions` InSpec audit resource to test properties of a all the AWS Traffic Mirror Sessions.
```

File: docs/resources/aws_ec2_traffic_mirror_sessions.md [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+
+## Syntax
+
+An `aws_ec2_traffic_mirror_sessions` resource block declares the tests for a all the  AWS Traffic Mirror Sessions
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_sessions.md:12]:
```suggestion
An `aws_ec2_traffic_mirror_sessions` resource block declares the tests for all the  AWS Traffic Mirror Sessions.
```

File: docs/resources/aws_ec2_traffic_mirror_sessions.md [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+end
+
+
+##### aws_\ec2\_traffic\_mirror\_sessions
+
+The ID of the AWS Traffic Mirror Session ID. This is in the format of `tms-` followed by 8 or 17 hexadecimal characters.
+This can be passed either as a string or as an `aws_ec2_traffic_mirror_sessions: 'value'` key-value entry in a hash.
+
+
+
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_sessions.md:25]:
```suggestion
```

File: docs/resources/aws_ec2_traffic_mirror_sessions.md [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+end
+
+##### Test that Traffic  mirror target includes a value
+describe aws_ec2_traffic_mirror_sessions do
+  its('traffic_mirror_target_id') { should include 'test-mirror-id' }
+end
+
+##### Test that description is present
+describe aws_ec2_traffic_mirror_sessions do
+  its('description') { should include 'test-lt' }
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_sessions.md:56]:
```suggestion
  its('description') { should include 'test-description' }
```

File: test/unit/resources/aws_ec2_traffic_mirror_session_test.rb [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+
+  def test_rejects_scalar_invalid_args
+    assert_raises(ArgumentError) { AWSEc2TrafficMirrorSession.new('dummy') }
+  end
+
+  def test_rejects_unrecognized_params
+    assert_raises(ArgumentError) { AWSEc2TrafficMirrorSession.new(rubbish: 9) }
+  end
+
+  def test_rejects_invalid_launch_template_endpoint_id
```
Comment (soumyo13) [inline test/unit/resources/aws_ec2_traffic_mirror_session_test.rb:18]: Change the method name

**Review by soumyo13 — CHANGES_REQUESTED** [review]
Please make the changes

File: docs/resources/aws_ec2_traffic_mirror_session.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+##### Test the packet_length of a Traffic mirror session
+    describe aws_ec2_traffic_mirror_session(traffic_mirror_session_id: 'test-traffic_mirror_session_id') do
+      its('packet_length') { should eq 1 }
+    end
+
+##### Test that a Traffic mirror session has the correct ID 
+    describe aws_ec2_traffic_mirror_session(traffic_mirror_session_id: 'test-traffic_mirror_session_id') do
+      its('traffic_mirror_session_id') { should eq 'test-lt' }
```
Comment (soumyo13) [inline docs/resources/aws_ec2_traffic_mirror_session.md:61]: its('traffic_mirror_session_id') { should eq 'test-traffic_mirror_session_id' }

**Review by soumyo13 — APPROVED** [review]
LGTM

**Review by IanMadd — APPROVED** [review]

**Pull request merged**